src/internal/pfsdb/repos.go

```diff
@@ -0,0 +1,363 @@
+func ListRepo(ctx context.Context, tx *pachsql.Tx) (*RepoIterator, error) {
+	return listRepo(ctx, tx, "", nil)
+}
+
+// ListRepoByIdxType is like ListRepo but only iterates over repo.type = repoType.
+func ListRepoByIdxType(ctx context.Context, tx *pachsql.Tx, repoType string) (*RepoIterator, error) {
+	return listRepo(ctx, tx, "type", repoType)
+}
+
+// ListRepoByIdxName is like ListRepo but only iterates over repo.name = name.
+func ListRepoByIdxName(ctx context.Context, tx *pachsql.Tx, repoName string) (*RepoIterator, error) {
+	return listRepo(ctx, tx, "name", repoName)
+}
+
+// ListRepo returns a RepoIterator that exposes a Next() function for retrieving *pfs.RepoInfo references.
+func listRepo(ctx context.Context, tx *pachsql.Tx, where string, whereVal interface{}) (*RepoIterator, error) {
+	limit := 100
+	page, err := listRepoPage(ctx, tx, limit, 0, where, whereVal)
+	if err != nil {
+		return nil, errors.Wrap(err, "list repos")
+	}
+	iter := &RepoIterator{
+		repos: page,
+		limit: limit,
+		tx:    tx,
+	}
+	if where != "" && whereVal != nil {
+		iter.where = where
+		iter.whereVal = whereVal
+	}
+	return iter, nil
+}
+
+func listRepoPage(ctx context.Context, tx *pachsql.Tx, limit, offset int, where string, whereVal interface{}) ([]repoRow, error) {
+	var page []repoRow
+	if where != "" && whereVal != nil {
+		if err := tx.SelectContext(ctx, &page,
+			fmt.Sprintf("%s WHERE repo.%s = $1 GROUP BY repo.id, project.name ORDER BY repo.id ASC LIMIT $2 OFFSET $3;", getRepoAndBranches, where),
+			whereVal, limit, offset); err != nil {
+			return nil, errors.Wrap(err, "could not get repo page")
+		}
+		return page, nil
+	}
+	err := tx.SelectContext(ctx, &page, getRepoAndBranches+"GROUP BY repo.id, project.name ORDER BY repo.id ASC LIMIT $1 OFFSET $2 ;", limit, offset)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get repo page")
+	}
+	return page, nil
+}
+
+// CreateRepo creates an entry in the pfs.repos table.
+func CreateRepo(ctx context.Context, tx *pachsql.Tx, repo *pfs.RepoInfo) error {
+	if repo.Repo.Type == "" {
+		repo.Repo.Type = "unknown"
+	}
+	if repo.Repo.Project == nil {
+		repo.Repo.Project = &pfs.Project{Name: "default"}
+	}
+	_, err := tx.ExecContext(ctx, "INSERT INTO pfs.repos (name, type, project_id, description) "+
+		"VALUES ($1, $2::pfs.repo_type, (SELECT id from core.projects where name=$3), $4);",
+		repo.Repo.Name, repo.Repo.Type, repo.Repo.Project.Name, repo.Description)
+	if err != nil && IsErrRepoAlreadyExists(err) {
+		return ErrRepoAlreadyExists{Project: repo.Repo.Project.Name, Name: repo.Repo.Name}
+	}
+	return errors.Wrap(err, "create repo")
+}
+
+// DeleteRepo deletes an entry in the pfs.repos table.
+func DeleteRepo(ctx context.Context, tx *pachsql.Tx, repoProject, repoName, repoType string) error {
+	result, err := tx.ExecContext(ctx, "DELETE FROM pfs.repos "+
+		"WHERE project_id=(SELECT id FROM core.projects WHERE name=$1) AND name=$2 AND type=$3;", repoProject, repoName, repoType)
```

Comment:
I'm partial to a `LIMIT 1` here, but that might be a little tin-foil-hat.

Reply:
Looks like this doesn't work: https://www.crunchydata.com/blog/simulating-update-or-delete-with-limit-in-postgres-ctes-to-the-rescue

If it helps, there is a unique constraint on this table so that (project.name, name, type) should be unique